Add Overfitting Test

RongLirr opened the pull request.

Commit:
Add test_overfit.py

RongLirr committed:
Merge pull request #1 from sign-language-processing/main

Merge pull request #5 from RongLirr/main

Commit:
Add matplotlib to dependencies

Commit:
Add dependencies

RongLirr committed:
Merge pull request #2 from sign-language-processing/main

Set PYTHONPATH for project and CAMDM

RongLirr committed:
Merge pull request #3 from sign-language-processing/main

Update test.yaml

Commit:
Add dependencies

Commit:
Fix import error in test_models.py

Commit:
Add sign-language-datasets Dependency

RongLirr committed:
Install Compatible NumPy

Install Compatible NumPy

Commit:
Add Compatible NumPy to dependencies

Commit:
Add Compatible tensorflow-datasets to dependencies

Commit:
Add Compatible tensorflow-datasets to dependencies

RongLirr commented:
<img width="1057" alt="图片" src="https://github.com/user-attachments/assets/a2b89dbf-866e-4279-bd16-e343a390dba7" />

I tried to manually define an older version of tensorflow-datasets in dependencies but it didn't work...Do you have any idea how I should deal with this error in pytest?

Commit:
Update overfit test

Commit:
Update overfit test

RongLirr commented:
The differences between the two samples are:

**Target Output (data)**
Sample 0: A linearly increasing sequence with small Gaussian noise.
Sample 1: A sine wave-based pattern (oscillating up and down) perturbed with noise.

**Input Condition (input_sequence)**
Sample 0: All zeros.
Sample 1: All ones.

**Target Mask (target_mask)**
Alternating masking patterns per sample (even or odd frames masked with 0)

The results of this test are:

out1 shape: torch.Size([1, 40, 178, 3]), out2 shape: torch.Size([1, 40, 178, 3])
Output L2 norm diff: 1.911246
Average keypoint error: 0.017711
Cosine distance: 0.000322
![overfit_loss_curve](https://github.com/user-attachments/assets/716e1a95-5185-487b-92da-eed423e3de9e)

AmitMY commented:
great test. nice loss graph. 
Please note the `lint` issues - you created much code and added many comments to `load_data` etc.
Pytest currently fails because of `tensorflow-datasets` - can you please pin it in `pyproject` to version [4.9.7](https://github.com/tensorflow/datasets/releases/tag/v4.9.7)?

Commit:
Update tensorflow-datasets Version

RongLirr commented:
> great test. nice loss graph. Please note the `lint` issues - you created much code and added many comments to `load_data` etc. Pytest currently fails because of `tensorflow-datasets` - can you please pin it in `pyproject` to version [4.9.7](https://github.com/tensorflow/datasets/releases/tag/v4.9.7)?

I’ve pinned tensorflow-datasets==4.9.7 in my pyproject.toml, but the same error still persists...

Update on Apr. 1: Fixed ✅

Commit:
Add normalization/denormailization to training pipeline

RongLirr commented:
Summary of Changes in `training.py`

### 1. **Normalization and Denormalization**
- **What I did:**  
  Added per-sample min-max normalization before training, and corresponding denormalization after inference.

- **Why:**  
  Normalization helps stabilize model training by ensuring that all pose values are within a consistent range (the loss stayed huge without this step). Denormalization restores the predicted poses to their original scale before exporting to `.pose` files.

### 2. **Confidence Calculation**
- **What I did:**  
  During export, each keypoint's velocity (frame-to-frame L2 distance) is computed. A confidence score is assigned using:  confidence = exp(-velocity)

- **Question:**  
  Is there a more suitable method for computing confidence values from keypoint to generate the pose? I'm highly unsure at this point.

### 3. **Pose Header Replacement**
- **What I did:**  
  Replaced the original pose header used in training with the dataset's true header during sample export.

- **Why:**  
  Previously exported `.pose` files failed to render properly and triggered the error:  
  `Offset is outside the bounds of the DataView`.

  This was due to a mismatch between the pose data and its header. Specifically, I used  
  `PoseHeader.read(BufferReader(buffer.read()))` to load a header directly from the **DGS corpus**, which defines **534 keypoints**, whereas my model only outputs **178 keypoints**. The viewer failed to parse the data array because of this mismatch.

Will update other changes / try to fix pylint issues tmrw

Commit:
Fix some of lint issues in training.py

Commit:
WIP:fFix some of lint issues

Commit:
Fix Enum mismatch in test: model_mean_type in diffuse()

Commit:
WIP:fFix some of lint issues

Commit:
WIP:fix some of lint issues

Commit:
WIP:fix more lint issues

Commit:
WIP:fix more lint issues

Commit:
fix(test_overfit): disable pylint false positive on F.cosine_similarity

Commit:
fix: replace self.loss_type with self.diffusion.loss_type to resolve pylint E1101

Commit:
fix more lint W/R issues

Commit:
Suppress pylint warnings for protected access and renamed arguments in training.py

Commit:
Suppress R0402: keep 'import torch.nn as nn' for clarity and PyTorch convention

Commit:
fix more lint W/R/C issues

Commit:
fix more lint W/R/C/E issues

RongLirr commented:
After quite a few rounds, the current code is now rated at **9.72/10** by `pylint`, but unfortunately still fails the Lint check due to a few remaining complexity-related warnings 😓. These include:

- Too many arguments or local variables in some functions (`R0913`, `R0914`)
- Too many instance attributes or statements in classes (`R0902`, `R0915`)

Most of these are flagged in core modules like `training.py`, `models.py`, and `gaussian_diffusion.py`. Since many of the affected classes and functions are **subclasses or adapted versions of components from CAMDM**, and now form the backbone of my training pipeline, I’m hesitant to refactor them too aggressively.

Separately, in `load_data.py`, I revised the **per-frame masking logic** (as mentioned in the issue I opened). Previously it relied on `pose.body.mask`, but I’ve now moved to directly compute the mask from the raw pose data as follows:

```python
target_mask = np.any(fluent_clip != 0, axis=(1, 2, 3))  # shape: [T]
```

Would love to hear your thoughts on both the Lint issue and the new masking logic. Thanks as always!

Commit:
Add safe globals and patch torch.load to fix weights-only loading issue
